propose-downstream: Detect identical patches

csomh opened the pull request:
The 'propose-downstream' command is using 'git format-patch' to generate
and update the patch files in the dist-git repository. These new patch
files will always be 'git-formatted', while existing patches might use
some different format and this will result an unnecessary update of the
patch-file.

In order to avoid this, attempt to massage the previously present
patch-files in the dist-git repo to a format which can be used with 'git
patch-id --stable', and tell if updating a patch-file can be skipped by
comparing the current and previous patch-id.

The code was tested on the f34 branch of glibc.

Signed-off-by: Hunor Csomortáni <csomh@redhat.com>

thrix commented:
/packit test

(testing new rawhide & f34 composes due to a know bug https://bugzilla.redhat.com/show_bug.cgi?id=1930965)

jpopelka commented on packit/git_utils.py:
```diff
@@ -109,3 +92,37 @@ def get_metadata_from_message(commit: git.Commit) -> Optional[dict]:
+    repl = r"\1"
+    patch = re.sub(pattern, repl, patch)
+
+    # Add missing 'diff --git' lines
```
Can the 2. and 3. blocks of code be swapped?
I'd spent less brainpower for switching from 'diff --git' to timestamps and back to 'diff --git'

csomh replied:
With a changed order the regexp pattern for the substitution to add missing diff lines would become more complicated, I consider reading regexp requiring more brainpower than a strange ordering of blocks :) Added a note that the order matters.

jpopelka commented on packit/patches.py:
```diff
@@ -456,6 +438,45 @@ def process_git_am_style_patches(
+        repo: git.Repo,
+    ) -> List[PatchMetadata]:
+        """
+        Check if the patches changed are indeed a change or can be reset.
```
Seeing this the first time I have no idea how is the output list different from the input one (`patch_list`).
I guess the output is just somehow filtered input, but I don't want to read (if I was not reviewing this PR) the code to discover it.

csomh replied:
Renamed the method and improved the docstring. Let me know if still requires improvement.

jpopelka commented on packit/patches.py:
```diff
@@ -456,6 +438,45 @@ def process_git_am_style_patches(
+                current_patch_id = repo.git.patch_id(istream=fp).split()[0]
+                logger.debug(f"Current patch-id: {current_patch_id}")
+
+            if current_patch_id != current_patch_id:
```
```suggestion
            if current_patch_id != prev_patch_id:
```

csomh replied:
Wow! Thanks!

jpopelka commented on packit/git_utils.py:
```diff
@@ -109,3 +92,37 @@ def get_metadata_from_message(commit: git.Commit) -> Optional[dict]:
     return None
+
+
+def git_patch_ish(patch: str) -> str:
```
Can we have some test for this?

csomh replied:
Tests added.

jpopelka reviewed (commented):
Just some details, looks OK otherwise.

TomasTomecek commented on packit/patches.py:
```diff
@@ -496,6 +517,9 @@ def create_patches(
                     patches, commits, destination, files_to_ignore
                 )
                 patch_list = self.process_git_am_style_patches(patch_list)
+                repo = is_git_repo(destination)
+                if repo:
+                    patch_list = self.check_patch_changes(patch_list, repo)
```
we should only do this for propose-downstream and make sure all patches are correctly created for `srpm`

csomh replied:
That's what the `if repo` part is doing. But to be honest, I'm not entierly sure I understand this whole code entierly well, it's super convoluted. I'll check one more time whether this is the right place to do this check of the patches.

TomasTomecek replied:
SRPM builds would definitely start failing if we filtered out all the patches :) so if the current code works for you in the glibc repo (and tests are passing), then the `if` condition should indeed be doing the trick

csomh replied:
Okay, so I moved undoing the identical patches to `api.sync_release()` after the `create_patches` call, which made checking for the git repo unnecessary.

It seems easier to understand what's happening this way.

Let me know if you think otherwise.

TomasTomecek reviewed (commented):
this indeed looks very promising and should give us the perfect mechanism for putting source-git patches downstream

thanks for figuring it out!

csomh commented on packit/git_utils.py:
```diff
@@ -109,3 +92,37 @@ def get_metadata_from_message(commit: git.Commit) -> Optional[dict]:
+    return patch
+
+
+def is_git_repo(destination: str) -> Optional[git.Repo]:
```
This was not needed after all.

csomh commented:
@jpopelka @TomasTomecek this is ready for another round of review. Main changes were:
- Added tests.
- Moved detecting identical patches out of `create_patches`, and it's called after `create_patches` in `sync_release()`. This way it doesn't interfere with patch creation for SRPMs.
- It seems that one patch is not handled correctly by this code. I've included this in the tests (see `weird-identical.patch`) and wrote about it in the test. Let me know if this needs more clarification. Atm, I cannot come up with a general approach to handle this particular case. Comparing patches where one side was seemingly randomly edited by hand is quite challenging.

csomh commented on tests/integration/test_source_git.py:
```diff
@@ -45,7 +45,6 @@
 def test_basic_local_update_without_patching(
     sourcegit_and_remote,
     distgit_and_remote,
-    mock_patching,
```
@lachmanfrantisek I'm not sure about the intention of this test. The name suggests, that this should be without patching, and so it seems to me that `mock_patching` can be removed. Am I misunderstanding something?

lachmanfrantisek replied:
You are right, it does not make sense to have it. (Maybe only the same calls with `.times(0)`..;)

csomh replied:
Thanks for the confirming it! :slightly_smiling_face:

thrix commented:
/packit test

TomasTomecek reviewed (approved):
LGTM, very nice work

csomh commented:
Thank you for the feedback @jpopelka and @TomasTomecek! They greatly improved this change.
